src/ipld/cid_hashmap.rs

```diff
@@ -0,0 +1,167 @@
+        self.v1_dagcbor_blake2b_hash_map.capacity() + self.fallback_hash_map.capacity()
+    }
+
+    /// Returns a reference to the value corresponding to the key.
+    pub fn get(&self, k: Cid) -> Option<&V> {
+        match k.try_into() {
+            Ok(CidVariant::V1DagCborBlake2b(bytes)) => self.v1_dagcbor_blake2b_hash_map.get(&bytes),
+            Err(()) => self.fallback_hash_map.get(&k),
+        }
+    }
+
+    /// Returns the number of elements in the map.
+    pub fn len(&self) -> usize {
+        self.v1_dagcbor_blake2b_hash_map.len() + self.fallback_hash_map.len()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use cid::{
+        multihash::{self, MultihashDigest},
+        Cid,
+    };
+    use fvm_ipld_encoding::DAG_CBOR;
+    use quickcheck_macros::quickcheck;
+
+    fn generate_hash_maps(cid_vector: Vec<(Cid, u64)>) -> (CidHashMap<u64>, HashMap<Cid, u64>) {
+        let mut cid_hash_map = CidHashMap::new();
+        let mut hash_map = HashMap::new();
+        for item in cid_vector.iter() {
+            cid_hash_map.insert(item.0, item.1);
+            hash_map.insert(item.0, item.1);
+
+            // Quickcheck does not reliably generate the DAG_CBOR/Blake2b variant of V1 CIDs; need to ensure we have enough samples of this variant in the map for testing, so generate this variant from the values in the key-value pairs.
+            let cid_v1 = Cid::new_v1(
+                DAG_CBOR,
+                multihash::Code::Blake2b256.digest(&item.1.to_be_bytes()),
+            );
+            cid_hash_map.insert(cid_v1, item.1);
+            hash_map.insert(cid_v1, item.1);
+        }
+        (cid_hash_map, hash_map)
+    }
+
+    #[quickcheck]
+    fn insert_new_key_is_none(cid_vector: Vec<(Cid, u64)>, cid: Cid, payload: u64) {
+        let (mut cid_hash_map, _) = generate_hash_maps(cid_vector);
+        // Quickcheck occasionally generates a key that is already present in the map, so remove it if it is present.
+        if cid_hash_map.contains_key(cid) {
+            cid_hash_map.remove(cid);
+        }
+        assert!(cid_hash_map.insert(cid, payload).is_none());
+    }
+
+    #[quickcheck]
+    fn insert_existing_key_is_some(cid_vector: Vec<(Cid, u64)>, cid: Cid, payload: u64) {
+        let (mut cid_hash_map, _) = generate_hash_maps(cid_vector);
+        cid_hash_map.insert(cid, payload);
+        assert!(cid_hash_map.insert(cid, payload).is_some());
+    }
+
+    #[quickcheck]
+    fn contains_key(cid_vector: Vec<(Cid, u64)>, cid: Cid, insert: bool) {
+        let (mut cid_hash_map, mut hash_map) = generate_hash_maps(cid_vector);
+        // Quickcheck rarely generates a key that is already present in the maps, so insert it with 50% probability to test `contains_key` with an equal distribution of results.
+        if insert {
+            cid_hash_map.insert(cid, 0);
+            hash_map.insert(cid, 0);
+        }
+        assert_eq!(cid_hash_map.contains_key(cid), hash_map.contains_key(&cid));
+    }
```

Comment:
I would have created two tests: One for when `contains` returns true, and one for when it returns false. But this is okay as well.